src/startup.ts

```diff
@@ -1,60 +1,106 @@
-import { getClient } from "./clients/TorrentClient.js";
-import { MatchMode } from "./constants.js";
 import { CrossSeedError } from "./errors.js";
-import { logger } from "./logger.js";
-import { getRuntimeConfig } from "./runtimeConfig.js";
-import { validateTorrentDir } from "./torrent.js";
+import { Label, logger } from "./logger.js";
+import { Action } from "./constants.js";
 import { validateTorznabUrls } from "./torznab.js";
-import { validateAction } from "./action.js";
+import { getClient } from "./clients/TorrentClient.js";
+import {
+	RuntimeConfig,
+	getRuntimeConfig,
+	setRuntimeConfig,
+} from "./runtimeConfig.js";
+import { inspect } from "util";
+import { stat, access } from "fs/promises";
+import { constants } from "fs";
+
+/**
+ * validates existence, permission, and that a path is a directory
+ * @param path string of path to validate
+ * @param optionName name of the configuration key
+ * @returns true if path exists and has required permission
+ */
+async function verifyPath(
+	path: string,
+	optionName: string,
+	permissions: number
+): Promise<boolean> {
+	try {
+		if ((await stat(path)).isDirectory()) {
+			await access(path, permissions);
+			return true;
+		}
+	} catch (error) {
+		if (error.code === "ENOENT") {
+			logger.error(
+				`\tYour ${optionName} "${path}" is not a valid directory on the filesystem.\n`
+			);
+		} else {
+			logger.error(
+				`\tYour ${optionName} "${path}" has invalid permissions.`
+			);
+		}
+	}
+	return false;
+}
+
+/**
+ * verifies the config paths provided against the filesystem
+ * @returns true (if paths are valid)
+ */
+async function checkConfigPaths(): Promise<void> {
+	const { action, linkDir, dataDirs, torrentDir, outputDir } =
+		getRuntimeConfig();
+	let pathFailure: number = 0;
+
+	if (!(await verifyPath(torrentDir, "torrentDir", constants.R_OK))) {
+		pathFailure++;
+	}
 
-function validateOptions() {
-	const {
-		action,
-		rtorrentRpcUrl,
-		qbittorrentUrl,
-		transmissionRpcUrl,
-		delugeRpcUrl,
-		dataDirs,
-		linkDir,
-		matchMode,
-		skipRecheck,
-	} = getRuntimeConfig();
 	if (
-		action === "inject" &&
-		!(
-			rtorrentRpcUrl ||
-			qbittorrentUrl ||
-			transmissionRpcUrl ||
-			delugeRpcUrl
-		)
+		action == Action.SAVE &&
+		!(await verifyPath(
+			outputDir,
+			"outputDir",
+			constants.R_OK | constants.W_OK
+		))
 	) {
-		throw new CrossSeedError(
-			"You need to specify --rtorrent-rpc-url, --transmission-rpc-url, --qbittorrent-url, or --deluge-rpc-url when using '-A inject'."
-		);
+		pathFailure++;
 	}
-	if ((dataDirs && !linkDir) || (!dataDirs && linkDir)) {
-		throw new CrossSeedError(
-			"Data based matching requires both --link-dir and --data-dirs"
-		);
+
+	if (
+		linkDir &&
+		!(await verifyPath(linkDir, "linkDir", constants.R_OK | constants.W_OK))
+	) {
+		pathFailure++;
+	}
```

Comment:
this is probably the bad code-golfy part of my brain talking, but we could abuse the `+=` operator's type coercion like so:

<img width="153" alt="image" src="https://github.com/cross-seed/cross-seed/assets/9119498/5ad1f1a3-16d3-4aac-b970-fc74ede66522">

Feel free to take it or leave it. It would probably look something like this:
```suggestion
	if (linkDir) {
		pathFailures += !(await verifyPath(linkDir, "linkDir", constants.R_OK | constants.W_OK));
	}
```

Reply:
`Operator '+=' cannot be applied to types 'number' and 'boolean'.ts(2365)`

Reply:
ok no problem